include/deal.II/base/vectorization.h

```diff
@@ -4812,6 +4938,78 @@ class VectorizedArray<float, 4>
 
 #endif // DOXYGEN
 
+/**
+ * horizontal_add() functions.
+ */
+
+#if DEAL_II_VECTORIZATION_WIDTH_IN_BITS >= 128 && defined(__SSE2__)
+inline double
+VectorizedArray<double, 2>::horizontal_add()
+{
+  __m128d t1 = _mm_unpackhi_pd(data, data);
+  __m128d t2 = _mm_add_pd(data, t1);
+  return _mm_cvtsd_f64(t2);
+}
+
+
+
+inline float
+VectorizedArray<float, 4>::horizontal_add()
+{
+  __m128 t1 = _mm_movehl_ps(data, data);
+  __m128 t2 = _mm_add_ps(data, t1);
+  __m128 t3 = _mm_shuffle_ps(t2, t2, 1);
+  __m128 t4 = _mm_add_ss(t2, t3);
+  return _mm_cvtss_f32(t4);
+}
+#endif
+
+
+
+#if DEAL_II_VECTORIZATION_WIDTH_IN_BITS >= 256 && defined(__AVX__)
+inline double
+VectorizedArray<double, 4>::horizontal_add()
+{
+  VectorizedArray<double, 2> t1;
+  t1.data = this->get_low() + this->get_high();
```

Comment:
@kronbichler Do we need `DEAL_II_COMPILER_USE_VECTOR_ARITHMETICS` here?

Reply:
In principle yes, but I believe all compilers support it nowadays. On the other hand, why not simply write the `_mm_add_pd(get_low(), get_high())` command here? gcc will map it back to `operator+` if I remember correctly.

Reply:
gcc does, do you also know about other compilers (clang)?

Reply:
That is good enough. Also, the main reason to do `operator+` over `_mm_add_ps` is to be able to fuse it with nearby multiplications into FMAs and other possible compound instructions, which is highly unlikely here. One has to go through the tree in terms of additions, I see no way around.

Reply:
Thanks for the explanation, done!